File: 1046.last-stone-weight.py
#
# @lc app=leetcode id=1046 lang=python3
#
# [1046] Last Stone Weight
#

# @lc code=start
import heapq


class Solution:
    def lastStoneWeight(self, stones: List[int]) -> int:
        stones = [-stone for stone in stones]
        heapq.heapify(stones)
        while len(stones) > 1:
            s1, s2 = heapq.heappop(stones), heapq.heappop(stones)
            if s1 != s2:
                heapq.heappush(stones, s1 - s2)
        return -stones[0] if stones else 0


# @lc code=end

File: RustSolutions/src/1046.last-stone-weight.rs
/*
 * @lc app=leetcode id=1046 lang=rust
 *
 * [1046] Last Stone Weight
 */

// @lc code=start
use std::collections::BinaryHeap;
use std::cmp::Reverse;

impl Solution {
    pub fn last_stone_weight(stones: Vec<i32>) -> i32 {
        let mut stoneheap = stones.iter().fold(
            BinaryHeap::with_capacity(stones.len()),
            |mut heap, &stone| {
                heap.push(Reverse(stone));
                heap
            },
        );
        while stoneheap.len() > 1 {
            let (Reverse(stone1), Reverse(stone2)) = (stoneheap.pop().unwrap(), stoneheap.pop().unwrap());
            if stone1 != stone2 {
                stoneheap.push(Reverse(stone1 - stone2));
            }
        }
        stoneheap.pop().unwrap_or(Reverse(0)).0

    }
}
// @lc code=end

